hal/src/stm32f2xx/sleep_hal.cpp

```diff
@@ -100,6 +117,97 @@ static int configGpioWakeupSource(const hal_wakeup_source_base_t* wakeupSources)
     return SYSTEM_ERROR_NONE;
 }
 
+static int configAnalogWakeupSource(const hal_wakeup_source_base_t* wakeupSources, uint8_t* configured) {
+    auto source = wakeupSources;
+    while (source) {
+        if (source->type == HAL_WAKEUP_SOURCE_TYPE_LPCOMP) {
+            uint8_t hysteresis = 24; // 20mV * (4095 / 3300)
+            auto analogWakeup = reinterpret_cast<const hal_wakeup_source_lpcomp_t*>(source);
+            *configured = true;
+            Hal_Pin_Info* pinMap = HAL_Pin_Map();
+            uint16_t currVol = hal_adc_read(analogWakeup->pin);
+            hal_adc_sleep(true, nullptr); // Suspend ADC
+            if (pinMap[analogWakeup->pin].pin_mode != AN_INPUT) {
+                HAL_Pin_Mode(analogWakeup->pin, AN_INPUT);
+            }
+            NVIC_DisableIRQ(ADC_IRQn);
+            NVIC_ClearPendingIRQ(ADC_IRQn);
+
+            RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+            /* ADC Common Init */
+            ADC_CommonInitTypeDef ADC_CommonInitStructure = {};
+            ADC_CommonInitStructure.ADC_Mode = ADC_Mode_Independent;
+            ADC_CommonInitStructure.ADC_Prescaler = ADC_Prescaler_Div2;
+            ADC_CommonInitStructure.ADC_DMAAccessMode = ADC_DMAAccessMode_Disabled;
+            ADC_CommonInitStructure.ADC_TwoSamplingDelay = ADC_TwoSamplingDelay_5Cycles;
+            ADC_CommonInit(&ADC_CommonInitStructure);
+
+            // ADC1 configuration
+            ADC_InitTypeDef ADC_InitStructure = {};
+            ADC_InitStructure.ADC_Resolution = ADC_Resolution_12b; // 12-bits
+            ADC_InitStructure.ADC_ScanConvMode = DISABLE;
+            ADC_InitStructure.ADC_ContinuousConvMode = ENABLE;
+            ADC_InitStructure.ADC_ExternalTrigConvEdge = ADC_ExternalTrigConvEdge_None;
+            ADC_InitStructure.ADC_ExternalTrigConv = ADC_ExternalTrigConv_T1_CC1;
+            ADC_InitStructure.ADC_DataAlign = ADC_DataAlign_Right;
+            ADC_InitStructure.ADC_NbrOfConversion = 1;
+            ADC_Init(ADC1, &ADC_InitStructure);
+
+            // ADC1 regular channel configuration
+            ADC_RegularChannelConfig(ADC1, pinMap[analogWakeup->pin].adc_channel, 1, ADC_SampleTime_480Cycles);
+
+            // Analog watchdog configuration
+            ADC_AnalogWatchdogCmd(ADC1, ADC_AnalogWatchdog_SingleRegEnable);
+            ADC_AnalogWatchdogSingleChannelConfig(ADC1, pinMap[analogWakeup->pin].adc_channel);
+            uint16_t th = (uint16_t)(analogWakeup->voltage * (4095 / 3300.0));
+            th &= 0x00000FFF;
+            switch (analogWakeup->trig) {
+                case HAL_SLEEP_LPCOMP_ABOVE: {
+                    ADC_AnalogWatchdogThresholdsConfig(ADC1, th, 0);
+                    break;
+                }
+                case HAL_SLEEP_LPCOMP_BELOW: {
+                    ADC_AnalogWatchdogThresholdsConfig(ADC1, 4095, th);
+                    break;
+                }
+                case HAL_SLEEP_LPCOMP_CROSS: {
+                    if (currVol >= th) {
```

Comment:
I'd probably remove this altogether. As long as we guard [th - XmV, th + XmV] area it should be fine just to setup the analog watchdog and rely on it to trigger once we are out of the protected range.

So, we'd be setting up the thresholds as:
```c++
ADC_AnalogWatchdogThresholdsConfig(ADC1, th + hysteresis, th - hysteresis);
```

Reply:
If doing as you proposed, the device may exit sleep mode immediately, as the `hysteresis` is 15mV usually and the current voltage is most likely out of the protected range.

Reply:
That would violate the behavior we have implemented for Gen3.

Reply:
Can you elaborate why? I'm pretty sure it works the same way on Gen 3.

Reply:
On Gen3, we just set a threshold the LPCOM and the hysteresis is fixed and achieved by hardware. If the current analog value if either above or below the threshold and we specify the interrupt type to be `CROSS`, then the Gen3 device can enter sleep, because the current analog value is not crossing the threshold. But for Gen2, the current value is out of the guarded range, the analog watchdog to set NVIC pending, thus it cannot enter sleep mode actually.

Reply:
Okay, now I see. You are right that if we are immediately out of the protected range we'll get an interrupt and what we want instead is to only get it when we cross the threshold :+1: